hoomd/md/pytest/test_custom_force.py

```diff
@@ -0,0 +1,326 @@
+            arrays.force[:] = forces
+            arrays.potential_energy[:] = potential
+
+
+class MyPeriodicFieldGPU(MyPeriodicFieldCPU):
+
+    def __init__(self, A, i, p, w):
+        super().__init__(A, i, p, w)
+
+    def set_forces(self, timestep):
+        with self._state.gpu_local_snapshot as snap, \
+                self.gpu_local_force_arrays as arrays:
+            forces, potential = self._evaluate_periodic(snap)
+            arrays.force[:] = cupy.array(forces)
+            arrays.potential_energy[:] = cupy.array(potential)
+
+
+@pytest.mark.parametrize("force_cls", [MyPeriodicFieldCPU, MyPeriodicFieldGPU],
+                         ids=lambda x: x.__name__)
+def test_compare_to_periodic(force_cls, simulation_factory,
+                             two_particle_snapshot_factory):
+    """Test hoomd external periodic compared to a python version."""
+    # sim with built-in force field
+    snap = two_particle_snapshot_factory()
+    sim = simulation_factory(snap)
+    periodic = md.external.field.Periodic()
+    periodic.params['A'] = dict(A=1, i=0, p=1, w=1)
+    nvt = md.methods.NVT(hoomd.filter.All(), kT=1, tau=1)
+    integrator = md.Integrator(dt=0.005, forces=[periodic], methods=[nvt])
+    sim.operations.integrator = integrator
+
+    # sim with custom but equivalent force field
+    snap2 = two_particle_snapshot_factory()
+    sim2 = simulation_factory(snap2)
+    _skip_if_cupy_not_imported_and_gpu_device(sim2)
+    periodic2 = force_cls(A=1, i=0, p=1, w=1)
+    nvt2 = md.methods.NVT(hoomd.filter.All(), kT=1, tau=1)
+    integrator2 = md.Integrator(dt=0.005, forces=[periodic2], methods=[nvt2])
+    sim2.operations.integrator = integrator2
+
+    # run simulations next to each other
+    sim.run(100)
+    exit_test = _try_running_sim(sim2, 100)
+    if exit_test:  # return if the sim can't be run
+        return
+
+    snap_end = sim.state.get_snapshot()
+    snap_end2 = sim2.state.get_snapshot()
+
+    # compare particle properties
+    if sim.device.communicator.rank == 0:
+        positions1 = snap_end.particles.position
+        positions2 = snap_end2.particles.position
+        npt.assert_allclose(positions1, positions2)
+
+        velocities1 = snap_end.particles.velocity
+        velocities2 = snap_end2.particles.velocity
+        npt.assert_allclose(velocities1, velocities2)
+
+    forces1 = integrator.forces[0].forces
+    forces2 = integrator2.forces[0].forces
+    if sim.device.communicator.rank == 0:
+        npt.assert_allclose(forces1, forces2)
+
+    energies1 = integrator.forces[0].energies
+    energies2 = integrator2.forces[0].energies
+    if sim.device.communicator.rank == 0:
+        npt.assert_allclose(energies1, energies2)
+
+    torques1 = integrator.forces[0].torques
+    torques2 = integrator2.forces[0].torques
+    if sim.device.communicator.rank == 0:
+        npt.assert_allclose(torques1, torques2)
+
+    assert integrator.forces[0].virials == integrator2.forces[0].virials
```

Comment:
When virials aren't computed because the proper flags aren't set, the `force.virials` property returns `None`. `npt.assert_allclose` can't compare two None values.

Reply:
Why does this not fail then? This should return an array of Booleans.

Reply:
In this particular test, accessing the virials property just returns `None` because I don't set the flags to compute the virials. I don't know whether that is the desired behavior or not.